### PR TITLE
Standalone query param

### PR DESCRIPTION
**Pull request by alexey-anufriev**
This pull request relates to or closes the following issues: #61 

I have verified that…
- [x] the commits in this pull request follow the [conventionalcommits](https://www.conventionalcommits.org) pattern
- [x] tests have been updated or added
- [x] documentation has been updated to reflect the changes made

### CODE REVIEW

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import { QueryParamController } from './query-param-controller.interface';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+import { selectValueAccessor } from '../accessors/util';
+import { QueryParam } from '../model/query-param';
+import { QueryParamGroup } from '../model/query-param-group';
+
+/**
+ * Binds a standalone {@link QueryParam} to a DOM element.
+ *
+ * This directive accepts the name of a {@link QueryParam} without having outer {@link QueryParamGroup}.
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param.directive.ts]: It doesn't accept the name of one, it accepts the object itself, so I think this should read "This directive accepts a {@link QueryParam} …"

File: projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<docs-item>
+    <p>
+        In case if there is no need to group multiple query params into the same <span apiDocsLink>QueryParamGroup</span>
```
Comment (Airblader) [inline projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html]: I would rephrase this to something like

> If you only need a single parameter, you can also bind a QueryParam directly without creating a QueryParamGroup.

File: projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+        (for example, to have a single changes detection mechanism or cross-params interaction, see here
+        <docs-link [page]="DocsPage.MODEL_CONFIGURATION" fragment="QueryParam-combineWith"></docs-link>)
+        then it is also possible to use <span apiDocsLink>QueryParam</span> in a standalone mode.
+    </p>
+
+    <p>
+        Here's an example of a standalone <span apiDocsLink>QueryParam</span>
```
Comment (Airblader) [inline projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html]: `:` at the end :-)

File: projects/ngqp-demo/src/app/shared/docs-page.pipes.ts [hunk 4]
```diff
@@ -21,6 +21,8 @@ export class DocsPageNamePipe implements PipeTransform {
                 return 'Working with the Model';
             case DocsPage.GLOBAL_CONFIGURATION:
                 return 'Global Configuration';
+            case DocsPage.STANDALONE_QUERY_PARAM:
+                return 'Standalone Query Param';
```
Comment (Airblader) [inline projects/ngqp-demo/src/app/shared/docs-page.pipes.ts]: "Query Param" is a name we only use to refer to the class (and then it should be QueryParam without space), but I'd just use "Standalone Parameters" here

File: projects/ngqp/core/src/lib/accessors/util.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import { ControlValueAccessor } from '@angular/forms';
+import { DefaultControlValueAccessorDirective } from './default-control-value-accessor.directive';
+import { NGQP_BUILT_IN_ACCESSORS } from './ngqp-accessors';
+
+/**
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/accessors/util.ts]: I know this is how it was before, but can we add `@internal` here to prevent it from being publicly documented?

File: projects/ngqp/core/src/lib/directives/query-param-controller.interface.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { ControlValueAccessor } from '@angular/forms';
+
+/**
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param-controller.interface.ts]: `@internal`

File: projects/ngqp/core/src/lib/directives/query-param-group.service.ts [hunk 7]
```diff
@@ -142,9 +142,9 @@ export class QueryParamGroupService implements OnDestroy {
     }
 
     /**
-     * Deregisters a {@link QueryParamNameDirective} directive by referencing its name.
+     * Unregisters a {@link QueryParamController} by referencing its name.
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param-group.service.ts]: I would actually like to retain the old name (deregister) as none is clearly better and the change is unrelated to the feature. I also somewhat prefer "deregister" because it has a stronger connotation of "it was registered before and now we remove that" :-)

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * {@link https://angular.io/api/forms/ControlValueAccessor}.
+ */
+@Directive({
+    selector: '[queryParam]',
+    providers: [QueryParamGroupService],
+})
+export class QueryParamDirective implements QueryParamController, OnChanges, OnDestroy  {
+
+    /**
+     * The name of the {@link QueryParam} inside its parent {@link QueryParamGroup}.
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param.directive.ts]: I think this entire comment is copy-pasted and wrong.

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+     * Note that this does not refer to the [parameter name]{@link QueryParam#urlParam}.
+     */
+    @Input('queryParam')
+    public queryParam: QueryParam<any>;
+
+    public readonly name = 'param';
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param.directive.ts]: `@internal`

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        this.valueAccessor = selectValueAccessor(valueAccessors);
+    }
+
+    /** @ignore */
+    public ngOnChanges(changes: SimpleChanges): void {
+        const nameChange = changes['queryParam'];
+
+        if (nameChange) {
+            if (!nameChange.firstChange) {
+                throw new Error('Standalone query param change is not allowed.');
```
Comment (Airblader) [inline projects/ngqp/core/src/lib/directives/query-param.directive.ts]: I'd like to rephrase this to

> Changing the QueryParam bound in standalone mode is currently not supported.

File: projects/ngqp/core/src/test/standalone-input-text.spec.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    public param: QueryParam<string>;
+
+    constructor(private qpb: QueryParamBuilder) {
+        this.param = qpb.stringParam('q');
+    }
+
+}
+
+describe('ngqp', () => {
```
Comment (Airblader) [inline projects/ngqp/core/src/test/standalone-input-text.spec.ts]: We should change this to make the test descriptions different from the other tests. Maybe just "ngqp standalone parameters". I'll be refactoring this later anyway.

File: projects/ngqp/core/src/test/standalone-textarea.spec.ts [hunk 12]
```diff
@@ -0,0 +1,118 @@
+import { Component } from '@angular/core';
```
Comment (Airblader) [inline projects/ngqp/core/src/test/standalone-textarea.spec.ts]: I think this test file doesn't add anything relevant, it just uses a different component. It certainly makes sense to test the different component types for which we have ControlValueAccessor implementations, but not in the scope of standalone params. I think we can just remove this file (unless I'm missing something?)

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts (hunk 1 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param.directive.ts]: Fixed

File: projects/ngqp/core/src/lib/accessors/util.ts (hunk 5 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/accessors/util.ts]: Added.

File: projects/ngqp/core/src/lib/directives/query-param-controller.interface.ts (hunk 6 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param-controller.interface.ts]: Added.

File: projects/ngqp/core/src/lib/directives/query-param-group.service.ts (hunk 7 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param-group.service.ts]: Ok

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts (hunk 8 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param.directive.ts]: Shame on me, fixed.

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts (hunk 9 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param.directive.ts]: Added.

File: projects/ngqp/core/src/lib/directives/query-param.directive.ts (hunk 10 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/lib/directives/query-param.directive.ts]: Agree.

File: projects/ngqp/core/src/test/standalone-input-text.spec.ts (hunk 11 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/test/standalone-input-text.spec.ts]: Done.

File: projects/ngqp/core/src/test/standalone-textarea.spec.ts (hunk 12 above)
Comment (alexey-anufriev) [reply projects/ngqp/core/src/test/standalone-textarea.spec.ts]: Deleted.

File: projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html (hunk 2 above)
Comment (alexey-anufriev) [reply projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html]: Done.

File: projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html (hunk 3 above)
Comment (alexey-anufriev) [reply projects/ngqp-demo/src/app/docs-items/standalone-query-param/standalone-query-param-docs.component.html]: Added `.`

File: projects/ngqp-demo/src/app/shared/docs-page.pipes.ts (hunk 4 above)
Comment (alexey-anufriev) [reply projects/ngqp-demo/src/app/shared/docs-page.pipes.ts]: Fixed.
